### ISSUE TITLE
Port twisted.internet.main to Python 3

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5973]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)| @itamarst reported|
|-|-|
|Trac ID|trac#5973|
|Type|enhancement|
|Created|2012-09-13 17:34:28Z|
|Branch|https://github.com/twisted/twisted/tree/imain-py3-5973-2|

`twisted.internet.main` should run on Python 3.

<details><summary>Searchable metadata</summary>

```
trac-id__5973 5973
type__enhancement enhancement
reporter__itamar itamar
priority__normal normal
milestone__Python_3_3_Minimal Python 3.3 Minimal
branch__branches_imain_py3_5973_2 branches/imain-py3-5973-2
branch_author__itamarst itamarst
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1347557668000000 1347557668000000
changetime__1347642168000000 1347642168000000
version__None None
owner__itamar itamar

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_1"></a>|@itamarst commented|
|-|-|

(In [35633]) Branching to 'imain-py3-5973'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_2"></a>|@itamarst set owner to @exarkun|
|-|-|

Should work now. 

Started, but haven't looked at: http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/imain-py3-5973

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Add future imports to `twisted/internet/main.py`
  1. Relying on `addCleanup` to put the original reactor back makes me nervous.  What code is going to run between the test method returning and the `addCleanup` functions running?  What if it tries to import `twisted.internet.reactor`?  It seems safer to me to restore `sys.modules['twisted.internet.reactor']` before the test method returns.  (In fact, I wrote a utility for doing this in [#5935](https://github.com/twisted/twisted/issues/5935) which I guess is basically ready to be merged.)
  1. I don't understand the try/except or the new comment in `test_alreadyInstalled`.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_4"></a>|@itamarst commented|
|-|-|

(In [35650]) Address review comments 1 and 3. Refs [#5973](https://github.com/twisted/twisted/issues/5973)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_5"></a>|@itamarst commented|
|-|-|

(In [35651]) Branching to 'imain-py3-5973-2'

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_6"></a>|@itamarst commented|
|-|-|

(In [35653]) Address review comment 2. Refs [#5973](https://github.com/twisted/twisted/issues/5973)

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_7"></a>|@itamarst set owner to @exarkun|
|-|-|

OK, ready for review again. Started new buildbot run:
http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/imain-py3-5973-2

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun set owner to @itamarst|
|-|-|

  1. Shouldn't `NoReactor.__enter__` also delete `twisted.internet.reactor` if it exists?  Also, I think `self._currentReactor` is synonymous with `self.modules["twisted.internet.reactor"]`.

Looks good otherwise, please merge when you're happy with that one point.

**Comment by twisted-trac** [comment]
|[<img alt="itamarst's avatar" src="https://avatars.githubusercontent.com/u/3266662?s=50" width="50" height="50">](https://github.com/itamarst)<a name="note_9"></a>|@itamarst set status to `closed`|
|-|-|

(In [35665]) Merge imain-py3-5973-2.

Author: itamar
Review: exarkun
Fixes: [#5973](https://github.com/twisted/twisted/issues/5973)

Port twisted.internet.main to Python 3.
